keps/sig-storage/20180515-svcacct-token-volumes.md

```diff
@@ -159,6 +161,55 @@ sources:
     audience: ca.istio.io
 ```
 
+### File Permission
+
+The secret projections are currently written with world readable (0644) file
+permissions. Given that file permissions are one of the oldest and most hardened
+isolation mechanisms on unix, this is not ideal. We would like to
+opportunistically restrict permissions for projected service account tokens as
+long we can show that they won’t break users if we are to migrate away from
+secrets to distribute service account credentials.
+
+#### Proposed heuristics
+
++   *Case 1*: The pod has an fsGroup set. We can set the file permission on the
+    token file to 0600 and let the fsGroup mechanism work as designed. It will
+    set the permissions to 0640, chown the token file to the fsGroup and start
+    the containers with a supplemental group that grants them access to the
+    token file. This works today.
++   *Case 2*: The pod’s containers declare the same runAsUser for all containers
+    (ephemeral containers are excluded) in the pod. We chown the token file to
+    the pod’s runAsUser to grant the containers access to the token. All
+    containers must have UID either specified in container security context or
+    inherited from pod security context. Preferred UIDs in container images are
+    ignored.
```

Comment:
fsGroup and runAsUser fields can be modified by the PodSecurityPolicy admission plugin, and by mutating webhook admission plugins.

The service account admission plugin that sets up the token volume [currently runs before PodSecurityPolicy](https://github.com/kubernetes/kubernetes/blob/release-1.18/pkg/kubeapiserver/options/plugins.go#L74-L79). If we want it to be informed by fsGroup/runAsUser settings applied by PSP, it would need to be moved after the PodSecurityPolicy admission plugin. That would mean when PSP sees the pod, it would not yet have an auto-injected secret or token volume mount, which could change which pod security policy is selected/applied to the pod.

If a mutating webhook modifies the pod, then the ServiceAccount admission plugin Admit() method will be reinvoked. Because the webhook could have injected additional sidecar containers with different (or missing) runAsUser settings, that could affect the calculated token volume permissions.